app/src/lib/stores/app-store.ts

```diff
@@ -2045,12 +2086,141 @@ export class AppStore extends TypedBaseStore<IAppState> {
     const workingDirectory = WorkingDirectoryStatus.fromFiles(updatedFiles)
 
     this.repositoryStateCache.updateChangesState(repository, () => ({
-      diff,
+      selection: {
+        ...changesState.selection,
```

Comment:
There's a subtle problem here I wanted to call out and see how we feel.

Because we're inside a callback, `changesState.selection` reverts to the union type `ChangesSelection` rather than the known `ChangesWorkingDirectorySelection` that we were looking at above.

You can see this by extracting out the `selection` to a local variable and stating that it should satisfy `ChangesWorkingDirectorySelection`:

```ts
    this.repositoryStateCache.updateChangesState(repository, () => {
      const selection: ChangesWorkingDirectorySelection = {
        ...changesState.selection,
        diff,
      }

      return {
        selection,
        workingDirectory,
      }
    })
```

Which produces the compiler error because the previous checks for `changesState.selection` are no longer valid:

```
Type '{ diff: IDiff; kind: ChangesSelectionKind.WorkingDirectory; selectedFileIDs: string[]; } | { diff: IDiff; kind: ChangesSelectionKind.Stash; selectedStashedFile: CommittedFileChange | null; selectedStashedFileDiff: ITextDiff | ... 4 more ... | null; }' is not assignable to type 'ChangesWorkingDirectorySelection'.
  Property 'selectedFileIDs' is missing in type '{ diff: IDiff; kind: ChangesSelectionKind.Stash; selectedStashedFile: CommittedFileChange | null; selectedStashedFileDiff: ITextDiff | IImageDiff | IBinaryDiff | ILargeTextDiff | IUnrenderableDiff | null; }' but required in type 'ChangesWorkingDirectorySelection'.ts(2322)
app-state.ts(535, 12): 'selectedFileIDs' is declared here.
```

The way to appease the compiler is to create the `selection` before invoking `updateChangesState` when we can assert it's a valid state, which is what I'd like to see here:

```ts
    const selection: ChangesWorkingDirectorySelection = {
      ...changesState.selection,
      diff,
    }

    this.repositoryStateCache.updateChangesState(repository, () => {
      return {
        selection,
        workingDirectory,
      }
    })
```

Thoughts?

Reply:
That is some weak typechecking, I am disappoint. Given that `changesState` is a `const` and `selection` is a readonly property containing only readonly members. I suspect this is just an type checker performance issue given that extracting `changesState.selection` itself into a const will not yield an error.

> There's a subtle problem here

Just to be super clear, there's not really a problem right? Only a weak checker? I'll apply the suggestion you made regardless, just want to be sure that you didn't actually find a case where this could cause an issue.

Reply:
@niik yep, it's an limitation with the typechecker because it's lacks the context we built up earlier in the function